primitives/numeric/src/lib.rs

```diff
@@ -0,0 +1,531 @@
+pub struct Numeric {
+    inner: Decimal,
+}
+
+/// Decimal number constrained by [`NumericSpec`].
+/// Should be used for storage only, all math operations should be performed on unconstrained decimal and then converted to constrained
+#[derive(
+    Serialize,
+    Deserialize,
+    Clone,
+    Copy,
+    Debug,
+    derive_more::Display,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Hash,
+    Encode,
+    Decode,
+)]
+#[repr(transparent)]
+#[serde(transparent)]
+pub struct NumericConstrained {
+    numeric: Numeric,
+}
+
+/// Define maximum precision and scale for given number.
+///
+/// E.g.
+///
+/// 3.1415 has a scale of 4 and a precision of 5
+#[derive(
+    Clone,
+    Copy,
+    Debug,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Default,
+    Hash,
+    SerializeDisplay,
+    DeserializeFromStr,
+    Encode,
+    Decode,
+    iroha_schema::IntoSchema,
+)]
+#[cfg_attr(
+    all(feature = "ffi_export", not(feature = "ffi_import")),
+    derive(iroha_ffi::FfiType)
+)]
+pub struct NumericSpec {
+    /// Count of decimal digits in the fractional part.
+    /// Currently only positive scale up to 28 decimal points is supported.
+    pub scale: Option<u32>,
+}
+
+/// Error occurred during creation of [`Numeric`]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, displaydoc::Display)]
+#[cfg_attr(feature = "std", derive(thiserror::Error))]
+pub enum NumericError {
+    /// Mantissa exceeds allowed range
+    MantissaTooLarge,
+    /// Scale exeeds allowed range
+    ScaleTooLarge,
+    /// Negative values are not allowed
+    Negative,
+    /// Malformed: expecting number with optional decimal point (10, 10.02)
+    Malformed,
+}
+
+/// Error occurred while checking if number satisfy given spec
+#[derive(Clone, Copy, Debug, displaydoc::Display)]
+#[cfg_attr(feature = "std", derive(thiserror::Error))]
+pub enum NumericSpecError {
+    /// Given number has scale higher than allowed by spec.
+    ScaleTooHigh,
+}
+
+/// Error occurred while checking if number satisfy given spec
+#[derive(Clone, Debug, displaydoc::Display)]
+#[cfg_attr(feature = "std", derive(thiserror::Error))]
+pub enum NumericSpecParseError {
+    /// String representation should start with Numeric
+    StartWithNumeric,
+    /// Numeric should be followed by optional scale wrapped in braces
+    WrappedInBraces,
+    /// Scale should be valid integer value: {_0}
+    InvalidScale(#[cfg_attr(feature = "std", source)] <u32 as FromStr>::Err),
+}
+
+impl Numeric {
+    /// Zero numeric value
+    pub const ZERO: Self = Self::new(0, 0);
+    /// One numeric value
+    pub const ONE: Self = Self::new(1, 0);
+    /// Maximal numeric value
+    pub const MAX: Self = Self {
+        inner: Decimal::MAX,
+    };
+
+    /// Create new numeric given mantissa and scale
+    ///
+    /// # Panics
+    /// Panics in cases where [`Self::try_new`] would return error.
+    #[inline]
+    pub const fn new(mantissa: u128, scale: u32) -> Self {
```

Comment:
To use for const, panic will be caught during compilation.

Reply:
> To use for const, panic will be caught during compilation.

interesting. It has never occurred to me that I can use `const fn` like this. TBH, does it need a `# Panics` in the documentation? I would think that `panic` is a runtime concept

Reply:
You can check this [example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d14ffe7fb8244f193665732f6075f54a).

I think `# Panics` won't hurt, because we cant guarantee that this function would be used **only** in const context.

Reply:
I'm not entirely convinced that we should have this function in our API. Can you give some examples of this precedent in other crates?

Reply:
`rust_decimal` has this approach as well, anyway we would use this method almost exclusively in tests (in other cases `Numeric` would be deserialized).

Reply:
it's a bit weird approach. Is the error to be handled or not? If we take this to extreme, we could make a function that panics for every that returns a `Result`. @DCNick3 your thoughts?

Reply:
Also i want to notice that `Result::unwrap` is not `const fn` so it wouldn't be possible to create constant `Numeric` values directly. 
We would have to use `match` which essentially would result into reinvention of `Numeric::new`

Reply:
I think it's pretty normal to have a `try_new` to handle the errors and a shorthand panicking `new` alongside it, pretty common in libraries. I see no problem with that, especially considering the `const fn` use-case.